src/src/Commands/GPIO.cpp

```diff
@@ -0,0 +1,385 @@
+
+  		String log = logPrefix + String(F(" : port#")) + String(event->Par1) + String(F(": set to ")) + String(state);
+  		addLog(LOG_LEVEL_INFO, log);
+  		SendStatusOnlyIfNeeded(event->Source, SEARCH_PIN_STATE, key, log, 0);
+  		return return_command_success();
+  	} else {
+      logErrorGpioOffline(logPrefix,event->Par1);
+      return return_command_failed();
+    }
+  } else {
+    logErrorGpioOutOfRange(logPrefix,event->Par1);
+    return return_command_failed();
+  }
+}
+
+void logErrorGpioOffline(String prefix, byte port)
+{
+  String log;
+  log = prefix + String(F(": port# ")) + String(port) + String(F(" is offline."));
+  addLog(LOG_LEVEL_ERROR, log);
+}
+
+void logErrorGpioOutOfRange(String prefix, byte port)
+{
+  String log;
+  log = prefix + String(F(" port#")) + String(port) + String(F(" is out of range"));
+  addLog(LOG_LEVEL_ERROR, log);
+}
+
+void logErrorGpioNotOutput(String prefix, byte port)
+{
+  String log;
+  log = prefix + String(F(" port#")) + String(port) + String(F(" is not an output port"));
+  addLog(LOG_LEVEL_ERROR, log);
+}
+
+void createAndSetPortStatus_Mode_State(uint32_t key, byte newMode, int8_t newState)
+{
+  // WARNING: operator [] creates an entry in the map if key does not exist
+
+  globalMapPortStatus[key].mode     = newMode;
+  globalMapPortStatus[key].command  = 1; //set to 1 in order to display the status in the PinStatus page
+
+  //only force events if state has changed
+  if (globalMapPortStatus[key].state != newState) {
+    globalMapPortStatus[key].state        = newState;
+    globalMapPortStatus[key].output       = newState;
+    globalMapPortStatus[key].forceEvent   = 1;
+    globalMapPortStatus[key].forceMonitor = 1;
+  }
+}
+
+bool getPluginIDAndPrefix(char selection, byte &pluginID, String &logPrefix)
+{
+  bool success = true;
+  switch(tolower(selection))
+  {
+    case 'g': //gpio
+      pluginID=PLUGIN_GPIO;
+      logPrefix=F("GPIO");
+      break;
+    case 'm': //mcp
+      pluginID=PLUGIN_MCP;
+      logPrefix=F("MCP");
+      break;
+    case 'p': //pcf
+      pluginID=PLUGIN_PCF;
+      logPrefix=F("PCF");
+      break;
+    default:
+      logPrefix=F("PluginID out of range. Error");
+      success=false;
+  }
+  return success;
+}
+
+bool getPluginIDAndPrefixAndType(char selection, byte &pluginID, String &logPrefix, byte &gpioTimerType)
+{
+  bool success = true;
+  switch(tolower(selection))
+  {
+    case 'l': //longpulse (gpio)
+      pluginID =PLUGIN_GPIO;
+      logPrefix=F("GPIO");
+      gpioTimerType = GPIO_TYPE_INTERNAL;
+      break;
+    case 'm': //mcplongpulse (mcp)
+      pluginID =PLUGIN_MCP;
+      logPrefix=F("MCP");
+      gpioTimerType = GPIO_TYPE_MCP;
+      break;
+    case 'p': //pcflongpulse (pcf)
+      pluginID =PLUGIN_PCF;
+      logPrefix=F("PCF");
+      gpioTimerType = GPIO_TYPE_PCF;
+      break;
+    default:
+      logPrefix=F("PluginID out of range. Error");
+      success=false;
+  }
+  return success;
+}
```

Comment:
I will review the whole code, but just to explain the idea bihind the check on the first letter is due to two things:
1) I was trying to introduce a way to reduce the comand size and therefore the rules size. 
With these changes the user can write `monitor,g,1` instead of `monitor,gpio,1`. In the future I was planning to introduce short codes for the commands so that the rules file is shorter and therefore faster to parse.

2) I can use the `switch `statement that is fast

Reply:
I totally understand the optimization of the switch by looking at the first character.
I've also done it on the command parsing self.
Maybe you can first make a list of the commands you like to have in this PR?
Maybe we can also just have them defined as a separate command entry?
Or just have a single function to parse the command string and return an enum (a bit the same idea as you did in this function), but now you have two of these functions and that can be reduced to 1 :)

Reply:
ok. I have removed `getPluginIDAndPrefixAndType `and collapsed to one function.